src/transformers/modeling_tf_utils.py

```diff
@@ -196,10 +196,13 @@ def hf_compute_loss(self, labels, logits):
             from_logits=True, reduction=tf.keras.losses.Reduction.NONE
         )
         # make sure only labels that are not equal to -100 affect the loss
-        active_loss = tf.not_equal(tf.reshape(labels, (-1,)), -100)
-        reduced_logits = tf.boolean_mask(tf.reshape(logits, (-1, shape_list(logits)[2])), active_loss)
-        labels = tf.boolean_mask(tf.reshape(labels, (-1,)), active_loss)
-        return loss_fn(labels, reduced_logits)
+        unmasked_loss = loss_fn(labels, logits)
+        loss_mask = tf.cast(labels != -100, dtype=unmasked_loss.dtype)
+        loss_denominator = tf.reduce_sum(loss_mask, axis=1)
+        # Masked positions will have a loss of NaN because -100 is not a valid label
```

Comment:
I don't know if it's documented, but any cross-entropy loss will yield `NaN` or `-inf` when the model's probability for the true label is 0. Since labels < 0 do not correspond to a valid category, the loss in those cases will always be one of those values.